src/compile.c

```diff
@@ -0,0 +1,435 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "c2mir/c2mir.h"
+#include "cache.h"
+#include "compile.h"
+#include "decode.h"
+#include "mir-gen.h"
+#include "mir.h"
+#include "riscv_private.h"
+#include "utils.h"
+
+/* Provide for c2mir to retrieve the generated code string. */
+typedef struct {
+    char *code;
+    size_t code_size;
+    size_t curr;
+} code_string_t;
+
+/* mir module */
+typedef struct {
+    MIR_context_t ctx;
+    struct c2mir_options *options;
+    uint8_t debug_level;
+    uint8_t optimize_level;
+} riscv_jit_t;
+
+typedef struct {
+    char *name;
+    void *func;
+} func_obj_t;
+
+#define SET_SIZE_BITS 10
+#define SET_SIZE 1 << SET_SIZE_BITS
+#define HASH(val) \
+    (((val) * (GOLDEN_RATIO_32)) >> (32 - SET_SIZE_BITS)) & ((SET_SIZE) -1)
+#define sys_icache_invalidate(addr, size) \
+    __builtin___clear_cache((char *) (addr), (char *) (addr) + (size));
+
+/*
+ *  The set consists of SET_SIZE buckets, with each bucket containing 32 slots.
```

Comment:
Why 32 slots?

Reply:
I just set a sufficiently large number to prevent the bucket from becoming full.

Reply:
Add more comments and mark it as configurable. We shall always distinguish between magic numbers and configurations.